### DISCUSSION TITLE
Drop Tracer() method from Span API

### DISCUSSION THREAD

**Discussion by YuriiBarninets** [open-telemetry/opentelemetry-go#1887] [General]
Having [Tracer()](https://github.com/open-telemetry/opentelemetry-go/blob/d20e7228302077549282f1c2002418c90f341204/trace/trace.go#L500-L502) method within Span API introduces a kind of confusion in terms of delegating calls to a new Tracer once TracerProvider was configured. 
For better understanding let's take a look at the following code:
```
package main

import (
	"context"
	"go.opentelemetry.io/otel"
	sdk "go.opentelemetry.io/otel/sdk/trace"
)

func main() {
	proxyTracer := otel.Tracer("Default Tracer")
	_, span := proxyTracer.Start(context.Background(), "Span A")
	spanTracer := span.Tracer();

	otel.SetTracerProvider(sdk.NewTracerProvider()) // set SDK TracerProvider

	proxyTracer.Start(context.Background(), "Span B") // SDK Span will be created
	spanTracer.Start(context.Background(), "Span C") // noopSpan will be created, even though SDK TracerProvider is installed
}
```
The problem is that `spanTracer := span.Tracer();` will return an instance of [`noopTracer`](https://github.com/open-telemetry/opentelemetry-go/blob/d20e7228302077549282f1c2002418c90f341204/trace/noop.go#L41) instead of an instance of internal  [`tracer`](https://github.com/open-telemetry/opentelemetry-go/blob/d20e7228302077549282f1c2002418c90f341204/internal/global/trace.go#L115-L124) that supports a possibility to delegate calls to an installed one when TracerProvider was configured.
As a result, after SDK TracerProvider was installed we still get a `noopSpan` upon this call `spanTracer.Start(context.Background(), "Span C")`.

I have checked `Span` interface of other techs and have not found `Tracer()` methods.
Here are links to other techs: [JavaScript Span API](https://github.com/open-telemetry/opentelemetry-js-api/blob/9b3f9c75a57190841b5f091933932386f63a62ac/src/trace/span.ts#L32), [Java Span API](https://github.com/open-telemetry/opentelemetry-java/blob/0b8131c79e47a8d6ad4f6e7d94d2f902238e3a9c/api/all/src/main/java/io/opentelemetry/api/trace/Span.java#L28), [PHP Span API](https://github.com/open-telemetry/opentelemetry-php/blob/c3bc66d9c68b07f0f4afe005c08bb97b79720bd9/api/Trace/Span.php#L9), [.NET Span API](https://github.com/open-telemetry/opentelemetry-dotnet/blob/6cdfcea67c4abeb563a7fc809bdcb1bab4dfe22a/src/OpenTelemetry.Api/Trace/TelemetrySpan.cs#L28).

My proposal is to remove [Tracer()](https://github.com/open-telemetry/opentelemetry-go/blob/d20e7228302077549282f1c2002418c90f341204/trace/trace.go#L500-L502) method from Span API or return the internal [`tracer`](https://github.com/open-telemetry/opentelemetry-go/blob/d20e7228302077549282f1c2002418c90f341204/internal/global/trace.go#L115-L124)  upon `span.Tracer();` call for [`noopSpan`](https://github.com/open-telemetry/opentelemetry-go/blob/d20e7228302077549282f1c2002418c90f341204/trace/noop.go#L78)

**Comment by MrAlias** [comment]
This is a bug. The no-op span needs to return the `Tracer` that created it instead of a generic no-op `Tracer` implementation.

https://github.com/open-telemetry/opentelemetry-go/blob/d20e7228302077549282f1c2002418c90f341204/trace/noop.go#L78

**Reply by YuriiBarninets** [reply]
You know, it depends on from which point of view you are looking at the Span creation process.
If you are looking from the point of [the internal tracer](https://github.com/open-telemetry/opentelemetry-go/blob/d20e7228302077549282f1c2002418c90f341204/internal/global/trace.go#L119-L124) implementation then the Span was created by [noopTracer](https://github.com/open-telemetry/opentelemetry-go/blob/d20e7228302077549282f1c2002418c90f341204/internal/global/trace.go#L146), since internal `Tracer` redirects call to `noopTracer`.
If you are looking from the perspective of API consumers then, of course, it is a bug, since Client does not know anything about the internal `Tracer` placeholder.

Let me formalize it, so for this code
```
...
_, noopSpan := internalTracer.Start(context.Background(), "Noop Span")
tracerFromSpan := noopSpan.Tracer();
assert(internalTracer == tracerFromSpan);
```
Client is expecting to get that `internalTracer` object is equal to `tracerFromSpan` object since `noopSpan` was created by `internalTracer`.

**Reply by MrAlias** [reply]
Tracking with https://github.com/open-telemetry/opentelemetry-go/issues/1892

**Reply by Aneurysm9** [reply]
I'm not sure that trying to fix this is the right approach.  I've always felt wary of the pattern I've seen some people use where they obtain a span from an incoming context and use that span's tracer to start another span.  Since a tracer should contain a resource identifying the instrumentation library and version it is likely that re-using a tracer from a span created by a different library will result in incorrect identification of the source of the span.  Removing the `Tracer()` method from the span will force instrumentation authors to obtain and utilize their own tracer, which should at the least have not incorrect information.

**Reply by MrAlias** [reply]
For some reason I though the `Tracer` method was a part of the specification, it is not. Without this requirement from the specification, I agree we should go in the direction of removing this method.

**Comment by PeterFeicht** [comment]
I would argue that the current implementation is the right thing to do. The `Tracer` method's documentation says it returns the tracer that created the span, which it does. As a client I wouldn't expect the `Tracer` method to suddenly return a different tracer just because I configured a tracer provider, because that doesn't change the tracer that created the span.

I don't know the use case that prompted adding the `Tracer` method to the `Span` interface. But one use case I can imagine is that of tracing a transaction through multiple spans from beginning to end. In that case, changing the tracer from a noop tracer to a non-noop tracer in the middle of the transaction seems surprising, because spans for the transaction will start somewhere in the middle.

I can understand someone wanting to switch to the newly configured tracer, and that's possible by using a tracer obtained via `otel.Tracer`. I also realize that this is probably a corner case that rarely anybody will encounter in real life, just wanted to add my two cents.
